### PR TITLE
Feat/create the cylinder primitive

### PR DESCRIPTION
**Pull request by theophile-jr**
Unfinite cylinder ("should") - handling axis parameter of course

### CODE REVIEW

File: include/Utils/ObjParser.hpp [hunk 1]
```diff
@@ -0,0 +1,16 @@
+/*
```
Comment (mathieurobert1) [inline include/Utils/ObjParser.hpp:1]: Why this file is here ?

File: .gitignore [hunk 2]
```diff
@@ -30,3 +30,6 @@
 *.exe
 *.out
 *.app
+
+
+output.ppm
```
Comment (Njord201) [inline .gitignore:35]:
```suggestion
*.ppm
```

File: include/Primitives/Cylinder.hpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+             *
+             */
+            Cylinder();
+
+            /**
+             * @brief Construct a new Cylinder object
+             *
+             * @param origin center of the cylinder
+             * @param radius of the cylinder
+             * @param height of the cylinder
```
Comment (Njord201) [inline include/Primitives/Cylinder.hpp:28]: The third parameter is axis

File: include/Primitives/Cylinder.hpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+             * @param ray vector3D
+             * @return Point3D
+             */
+            Math::Point3D hitPoint(const Raytracer::Ray& ray) const override;
+
+            /**
+             * @brief set origin of cylinder
+             *
+             * @param hitPoint Point3D
+             * @return Vector3D
```
Comment (Njord201) [inline include/Primitives/Cylinder.hpp:50]: The function return nothing

File: include/Primitives/Cylinder.hpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+             */
+            void setOrigin(const Math::Point3D& origin);
+
+            /**
+             * @brief set axis of cylinder
+             *
+             * @return Point3D
```
Comment (Njord201) [inline include/Primitives/Cylinder.hpp:57]: The function return nothing

File: include/Primitives/Cylinder.hpp [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+             */
+            void setAxis(double axis);
+
+            /**
+             * @brief set radius of cylinder
+             *
+             * @return double
```
Comment (Njord201) [inline include/Primitives/Cylinder.hpp:64]: The function return nothing

File: src/Core/Parser/Scene.cpp [hunk 7]
```diff
@@ -132,6 +132,72 @@ int Raytracer::Scene::_parsePrimitiveSetting(const libconfig::Setting &primitive
...
+            if (axisType == "X") {
+                newCylinder->setAxis(Primitive::Axis::X);
+            } else if (axisType == "Y") {
+                newCylinder->setAxis(Primitive::Axis::Y);
+            } else if (axisType == "Z") {
+                newCylinder->setAxis(Primitive::Axis::Z);
+            } else {
+                throw ParserException("Wrong Axis for plane");
+            }
+            this->_primitives.add(newCylinder);
```
Comment (mathieurobert1) [inline src/Core/Parser/Scene.cpp:162]: Add also a translation vector as the camera, sphere or lights

File: include/Primitives/Cylinder.hpp [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+             */
+            void setRadius(double radius);
```
Comment (Njord201) [inline include/Primitives/Cylinder.hpp:66]: Describe parameter radius in documentation

File: include/Primitives/Cylinder.hpp [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+             * 
+             * @param axis The axis of cylinder Object to set
+             */
+            void setAxis(const Primitive::Axis &axis);
+
+        private:
+
+            Math::Point3D _origin;
+            double _radius;
+            Primitive::Axis                         _axis;
```
Comment (Njord201) [inline include/Primitives/Cylinder.hpp:88]: Keep logic of /t

File: include/Primitives/Plane.hpp [hunk 10]
```diff
@@ -33,7 +36,7 @@ namespace Primitive {
              * @param axis Axis of the plane
              * @param position offset on axis
              */
-            Plane(Primitive::Axis axis, double position);
+            Plane(Primitive::Axis axis, double position, std::shared_ptr<Material::IMaterial> material);
```
Comment (Njord201) [inline include/Primitives/Plane.hpp:39]: Describe parameter material in documentation

**Review by mathieurobert1 — CHANGES_REQUESTED** [review]
Some additions and deletions to do.
And are you sure you are up to date with the main ? Because I think some files are missing.

**Review by Njord201 — CHANGES_REQUESTED** [review]
See previous messages

File: include/Primitives/Plane.hpp (hunk 10 above)
Comment (theophile-jr) [reply include/Primitives/Plane.hpp:39]: it's from the merge of another branch

**Review by Njord201 — CHANGES_REQUESTED** [review]
Just a few requests: before merging, don't forget to merge the main branch. But good job! :1st_place_medal :

File: include/Primitives/Cylinder.hpp [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+            /**
+             * @brief Set the Axis object
+             * 
+             * @param axis The axis of cylinder Object to set
+             */
+            void setAxis(const Primitive::Axis &axis);
+
+            /**
+             * @brief Get the Material object
+             * @return Material of cylinder
```
Comment (Njord201) [inline include/Primitives/Cylinder.hpp:85]: Add an empty line

File: include/Primitives/Cylinder.hpp [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+             */
+            std::shared_ptr<Material::IMaterial> getMaterial() const;
+
+            /**
+             * @brief Set the Material object
+             * @param material Material of cylinder
```
Comment (Njord201) [inline include/Primitives/Cylinder.hpp:91]: Add an empty line

File: src/Plugins/Primitives/Cylinder/Cylinder.cpp [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+void Primitive::Cylinder::setRadius(double radius)
+{
+    this->_radius = radius;
+}
+
+Math::Point3D Primitive::Cylinder::computeColor(const Math::Point3D& hitPoint, const Light::LightsContainer& lights) const
+{
+    Math::Vector3D cylinderNormaml(hitPoint.x() - _origin.x(), hitPoint.y() - _origin.y(), hitPoint.z() - _origin.z());
```
Comment (mathieurobert1) [inline src/Plugins/Primitives/Cylinder/Cylinder.cpp:88]: rename cylinderNormaml by cylinderNormal

File: src/Plugins/Primitives/Cylinder/Makefile
Comment (Njord201) [inline src/Plugins/Primitives/Cylinder/Makefile:1]: Look at the other Makefile to make a nice Makefile

**Review by mathieurobert1 — CHANGES_REQUESTED** [review]

**Review by Njord201 — CHANGES_REQUESTED** [review]

File: src/Plugins/Primitives/Cylinder/Cylinder.cpp [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+Math::Point3D Primitive::Cylinder::computeColor(const Math::Point3D& hitPoint, const Light::LightsContainer& lights) const
+{
+    Math::Vector3D cylinderNormal(hitPoint.x() - _origin.x(), hitPoint.y() - _origin.y(), hitPoint.z() - _origin.z());
+
+    if (_material->getType() == Material::MaterialType::FlatColor) {
+        std::shared_ptr<FlatColor> cylinderFlatColor = std::dynamic_pointer_cast<FlatColor>(getMaterial());
+        return lights.computeColor(cylinderNormal, hitPoint, Math::Point3D(cylinderFlatColor->getR(), cylinderFlatColor->getG(), cylinderFlatColor->getB()));
+    }
+    std::cout << "material not handle in cylinder" << std::endl;
+    return Math::Point3D(0,0,0);
```
Comment (Njord201) [inline src/Plugins/Primitives/Cylinder/Cylinder.cpp:95]: Return VOID_COLOR macro

**Review by Njord201 — CHANGES_REQUESTED** [review]

**Review by Njord201 — APPROVED** [review]

**Review by mathieurobert1 — APPROVED** [review]

**Comment by theophile-jr** [comment]
For anyone reading this pull request, they voluntarly tried to make MANY request changes.
image

**Pull request merged**